DAOS-504 common: Allow delayed tse_task scheduling

liw opened the pull request:
Add a couple of new tse API methods for delayed task scheduling:

  - tse_task_schedule_with_delay
  - tse_task_reinit_with_delay

These will be used to implement backoffs when restarting TXs, retrying
rsvc operations, etc.

Signed-off-by: Li Wei <wei.g.li@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/1/execution/node/202/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/1/execution/node/270/log

daosbuild1 commented:
Test stage Build on CentOS 7 debug completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/1/execution/node/303/log

daosbuild1 commented:
Test stage Build on CentOS 7 release completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/1/execution/node/306/log

daosbuild1 commented:
Test stage Build on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/1/execution/node/301/log

daosbuild1 commented:
Test stage Build on Ubuntu 20.04 with Clang completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/1/execution/node/261/log

daosbuild1 commented:
Test stage Build on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/1/execution/node/276/log

daosbuild1 commented:
Test stage Build on Leap 15 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/1/execution/node/274/log

mchaarawi commented on src/common/tse.c:
```diff
@@ -925,14 +935,49 @@ tse_task_create(tse_task_func_t task_func, tse_sched_t *sched, void *priv,
+
 int
-tse_task_schedule(tse_task_t *task, bool instant)
+tse_task_schedule(tse_task_t *task, bool instant, uint64_t delay)
```
this is a public API and will break existing libraries using it (same for reinit).
maybe create a new API tse_task_set_delay() that delays a task for x seconds after being scheduled?

liw replied:
Sounds good to me. Will do.

liw replied:
I added tse_task_schedule_with_delay and tse_task_reinit_with_delay instead, because a tse_task_set_delay could not be called at certain times (e.g., after the task has been put in the sleeping queue sorted by wakeup_time).

mchaarawi commented on src/common/tse.c:
```diff
@@ -505,22 +507,29 @@ tse_task_complete_callback(tse_task_t *task)
-				      dtp_list) {
+	d_list_for_each_entry_safe(dtp, tmp, &dsp->dsp_sleeping_list,
+				   dtp_list) {
+		if (dtp->dtp_wakeup_time > now)
```
I didn't find where you set the wakeup_time for a task.

liw replied:
Yes, haven't added that. Will be in the new set_delay function.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/2/execution/node/999/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/3/execution/node/999/log

liw commented:
Rebased to the latest master to resolve conflicts with the new tse_task_reset. Fixed an issue where I was not setting dtp_wakeup_time to 0 in a few places.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/4/execution/node/970/log

mchaarawi commented on src/include/daos/common.h:
```diff
@@ -202,6 +202,15 @@ daos_getntime_coarse(void)
 }
 
+static inline uint64_t
+daos_getutime(void)
```
minor: but maybe add this to gurt?

liw replied:
Yeah, I thought about the same thing. There are already daos_get_ntime, daos_getntime_coarse, daos_gettime_coarse in this file though (look immediately above and below).

mchaarawi commented on src/include/daos/tse.h:
```diff
@@ -38,7 +39,7 @@
 #define TSE_PRIV_SIZE		1016
 /* tse_task arguments max length */
-#define TSE_TASK_ARG_LEN		888
+#define TSE_TASK_ARG_LEN		880
```
was there a reason for this?

liw replied:
Yes, for the new tse_task_private.dtp_wakeup_time.

liw commented:
@wangdi1, any chance you could review this PR? Or suggest another reviewer?

liw commented:
Merely merge with master; no changes.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3729/5/display/redirect

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/5/execution/node/1106/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/6/execution/node/292/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/6/execution/node/259/log

daosbuild1 commented:
Test stage Build on CentOS 7 debug completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/6/execution/node/359/log

daosbuild1 commented:
Test stage Build on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/6/execution/node/307/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-3729/7/execution/node/1024/log

daosbuild1 commented:
Test stage Functional_Hardware_Large completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3729/7/display/redirect

liw commented:
CI failures
  - daos_test -V -DER_INVAL: DAOS-6022
  - daos_test rebuild ec vos_byte2blkcnt: DAOS-6033
